packages/cli/src/cmds/validator/keymanager/decryptKeystoreDefinitions/index.ts

```diff
@@ -0,0 +1,105 @@
+import {spawn, ModuleThread, Pool, QueuedTask, Worker} from "@chainsafe/threads";
+import {SignerLocal, SignerType} from "@lodestar/validator";
+import bls from "@chainsafe/bls";
+import {LocalKeystoreDefinition} from "../interface.js";
+import {clearKeystoreCache, loadKeystoreCache, writeKeystoreCache} from "../keystoreCache.js";
+import {lockFilepath, unlockFilepath} from "../../../../util/lockfile.js";
+import {defaultPoolSize} from "./poolSize.js";
+import {DecryptKeystoreWorkerAPI, KeystoreDecryptOptions} from "./types.js";
+
+/**
+ * Decrypt keystore definitions using a threadpool
+ */
+export async function decryptKeystoreDefinitions(
+  keystoreDefinitions: LocalKeystoreDefinition[],
+  opts: KeystoreDecryptOptions
+): Promise<SignerLocal[]> {
+  if (opts.cacheFilePath) {
+    try {
+      const signers = await loadKeystoreCache(opts.cacheFilePath, keystoreDefinitions);
+      if (opts?.onDecrypt) {
+        opts?.onDecrypt(signers.length - 1);
+      }
+      opts.logger.debug("Loaded keystores via keystore cache");
+      return signers;
+    } catch {
+      // Some error loading the cache, ignore and invalidate cache
+      await clearKeystoreCache(opts.cacheFilePath);
+    }
+  }
+
+  const signers: SignerLocal[] = [];
+  const passwords: string[] = [];
+  const tasks: QueuedTask<ModuleThread<DecryptKeystoreWorkerAPI>, Uint8Array>[] = [];
+  const errors: Error[] = [];
+  const pool = Pool(
+    () =>
+      spawn<DecryptKeystoreWorkerAPI>(new Worker("./worker.js"), {
+        // A Lodestar Node may do very expensive task at start blocking the event loop and causing
+        // the initialization to timeout. The number below is big enough to almost disable the timeout
+        timeout: 5 * 60 * 1000,
+      }),
+    defaultPoolSize
+  );
+  for (const [index, definition] of keystoreDefinitions.entries()) {
+    try {
+      lockFilepath(definition.keystorePath);
+    } catch (e) {
+      if (opts.force) {
+        opts.logger.warn("Keystore forcefully loaded even though lockfile could not be acquired", {
+          path: definition.keystorePath,
+        });
+      } else {
+        throw e;
+      }
+    }
+
+    const task = pool.queue((thread) => thread.decryptKeystoreDefinition(definition, Boolean(opts.force)));
+    tasks.push(task);
+    task
+      .then((secretKeyBytes: Uint8Array) => {
+        const signer: SignerLocal = {
+          type: SignerType.Local,
+          secretKey: bls.SecretKey.fromBytes(secretKeyBytes),
+        };
+
+        signers.push(signer);
+        passwords.push(definition.password);
```

Comment:
> keystore cache does not seem to work, the file is created after import but on next startup is removed and keys are loaded from scratch

The issue is that the order of signers is non deterministic due to the fact that we run decryption in parallel and push them to the array asynchronously.

```
Keystore index: 5
Keystore index: 10
Keystore index: 4
Keystore index: 2
Keystore index: 11
Keystore index: 9
Keystore index: 0
Keystore index: 1
Keystore index: 8
Keystore index: 7
Keystore index: 6
Keystore index: 3
```

good news is, we have the index and can use that to ensure the order, updating the code as below fixes the issue

```suggestion
        signers[index] = signer;
        passwords[index] = definition.password;
```